# MNIST Neural Network

In [1]:
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'

In [2]:
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data

In [3]:
mnist = input_data.read_data_sets("/tmp/data/",one_hot=True)

Successfully downloaded train-images-idx3-ubyte.gz 9912422 bytes.
Extracting /tmp/data/train-images-idx3-ubyte.gz
Successfully downloaded train-labels-idx1-ubyte.gz 28881 bytes.
Extracting /tmp/data/train-labels-idx1-ubyte.gz
Successfully downloaded t10k-images-idx3-ubyte.gz 1648877 bytes.
Extracting /tmp/data/t10k-images-idx3-ubyte.gz
Successfully downloaded t10k-labels-idx1-ubyte.gz 4542 bytes.
Extracting /tmp/data/t10k-labels-idx1-ubyte.gz


# MNIST Dataset

We load the MNIST dataset that contains the dataset of handwritten digits saved as images.
Now we will train a neural network based upon these images and labels.

In [11]:
n_nodes_hl1 = 500
n_nodes_hl2 = 500
n_nodes_hl3 = 500

n_classes = 10

batch_size = 100

n_nodes_hl1,hl2,hl3 are the nodes in every hidden layer.<br>
n_classes are the total number of classes or the types of images.<br>
There are 10 types of images containing numbers 0 - 9.<br>
batch_size is a regular integer that contains the number of images to be trained every epoch.

In [12]:
x = tf.placeholder('float',[None,784]) # 784 is 28*28 ,i.e., the size of mnist images
y = tf.placeholder('float')

In [13]:
def neural_network_model(data):

    # biases are added so that the some neurons get fired even when input_data is 0
    hidden_1_layer = {'weights':tf.Variable(tf.random_normal([784,n_nodes_hl1])),'biases':tf.Variable(tf.random_normal([n_nodes_hl1]))}
    hidden_2_layer = {'weights':tf.Variable(tf.random_normal([n_nodes_hl1,n_nodes_hl2])),'biases':tf.Variable(tf.random_normal([n_nodes_hl2]))}
    hidden_3_layer = {'weights':tf.Variable(tf.random_normal([n_nodes_hl2,n_nodes_hl3])),'biases':tf.Variable(tf.random_normal([n_nodes_hl3]))}

    output_layer = {'weights':tf.Variable(tf.random_normal([n_nodes_hl3,n_classes])),
    'biases':tf.Variable(tf.random_normal([n_classes]))}


    # (input_data * weights) + biases

    l1 = tf.add(tf.matmul(data,hidden_1_layer['weights']) , hidden_1_layer['biases'])
    l1 = tf.nn.relu(l1)	# activation func

    l2 = tf.add(tf.matmul(l1,hidden_2_layer['weights']) , hidden_2_layer['biases'])
    l2 = tf.nn.relu(l2)	# activation func

    l3 = tf.add(tf.matmul(l2,hidden_3_layer['weights']) , hidden_3_layer['biases'])
    l3 = tf.nn.relu(l3)	# activation func

    output = tf.matmul(l3,output_layer['weights']) + output_layer['biases'] 

    return output

RELU or Rectifier Linear is an activation function.<br>
relu(x)==>max(0,x)

In [14]:
def train_neural_network(x):
    prediction = neural_network_model(x)
    cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits=prediction,labels=y))
    # softmax_cross_entropy_with_logits ==> for changing weights
    # we wanna minimize the difference

    # AdamOptimizer optionally has a learning_reate : 0.0001
    optimizer = tf.train.AdamOptimizer().minimize(cost)

    hm_epochs = 5 # cycles of feed forward + back

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer()) 

        for epoch in range(hm_epochs):
            epoch_loss = 0
            for _ in range(int(mnist.train.num_examples/batch_size)):
                epoch_x,epoch_y = mnist.train.next_batch(batch_size) 
                _,c = sess.run([optimizer, cost], feed_dict = {x: epoch_x, y: epoch_y})
                epoch_loss += c
            print('Epoch',epoch,'completed out of',hm_epochs,' loss:',epoch_loss)


        correct = tf.equal(tf.argmax(prediction,1),tf.argmax(y,1))
        accuracy = tf.reduce_mean(tf.cast(correct,'float'))
        # cast changes the datatype in tensorflow, in this case it will be changed to float type.
        print('Accuracy: ',accuracy.eval({x:mnist.test.images,y:mnist.test.labels}))

# Finally

Finally we will run the Neural Network to get results.<br>

In [15]:
if __name__ == "__main__":
    train_neural_network(x)

Epoch 0 completed out of 5  loss: 1593870.59964
Epoch 1 completed out of 5  loss: 375178.777066
Epoch 2 completed out of 5  loss: 204155.209455
Epoch 3 completed out of 5  loss: 122060.331427
Epoch 4 completed out of 5  loss: 73701.5975397
Accuracy:  0.9423


We finally have the accuracy attained by the Neural Network<br>It comes out to be: ~94 %